Fix 329

hiker opened the pull request:
This patch does not fix _all_ global variables (as in: variables that e.g. dynamo0p3 sets in psyGen), but it should fix all access related global variables. Since the patch is already reasonable big, I think it's better to limit the scope to access variables for now, and leave some of the other global variables for a future patch.

A side effect of this patch was that I renamed some of the global variables to have a GH_ in front of it in dynamo0p3 (that simplified my incremental work when I had some variables being the api-specific name, some the new api-independent access type), a future patch should remove them all similar to this patch removing FIELD_ACCESS_MAP , MAPPING_REDUCTIONS etc.

As discussed previously I added the access_type to a new directory '.../psyclone/core' (with tests in psyclone/core/tests), as a simple start to split those huge files into smaller ones. The idea of having tests under core/tests is that it is easy if you work in core to only run the tests closely related to that directory. Am happy to move that directory (there would be some case to be made to have it in .../src/psyclone/tests/core/ - mostly the usage for the conftest.py file under tests).

I also moved some of the constant lists (e.g. READ_ACCESSES) as functions into AccessType.

It also fixes #355.

arporter commented:
In case I've missed it, @TeranIvy - is the MO environment OK if we gain a dependence on enum?

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8102,25 +8112,24 @@ def unique_fs_names(self):
         are modified. If a kernel writes to more than one argument then
         that requiring the largest iteration space is selected.
 
-        :param dict mapping: un-used argument. Retained for consistency
```
Whoop!

arporter commented on src/psyclone/gocean0p1.py:
```diff
@@ -51,8 +52,20 @@ class GODescriptor(Descriptor):
+    on
+    :param dict stencil: type of stencil access for this \
+    argument. Defaults to None if the argument is not supplied.
+    :param string mesh: which mesh this argument is on. Defaults to \
```
s/string/str/

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -203,7 +199,7 @@ def args_filter(arg_list, arg_types=None, arg_accesses=None, arg_meshes=None,
     :type arg_types: list of str
     :param arg_accesses: List of access types that arguments must have
-    :type arg_accesses: list of str
+    :type arg_accesses: List of :py:class:`psyclone.core.access_type`.
```
s/access_type/AccessType/ I think?

hiker replied:
We were both wrong, as far as I can tell the right way should be: psyclone.core.access_type.AccessType.

Is there actually a way of testing this? I remembered that I saw a doxyfile in the past, but that has been reverted.

arporter replied:
I think you'd have to modify the documentation source files to say that you wanted Sphinx to generate the docs for this class. One day we will have an API reference guide and we would check in there that it was generated OK.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2278,13 +2268,16 @@ def view(self, indent=0):
         '''Return the name of all scalars within this region that require a
-        reduction of type reduction_type. Returned names will be unique. '''
+        reduction of type reduction_type. Returned names will be unique.
+        :param: reduction_type
```
Please complete this line documenting `reduction_type` :-)

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -2278,13 +2268,16 @@ def view(self, indent=0):
+        :param: reduction_type
+        :type reduction_type: :py:class:`psyclone.core.AccessType`
+        '''
+        print("Called")
```
and remove this line ;-)

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3124,26 +3117,24 @@ def __str__(self):
         return False
 
-    def unique_modified_args(self, mapping, arg_type):
+    def unique_modified_args(self, arg_type):
```
Please document the interface to this routine.

hiker replied:
The risk of "you touch it, you document it", since I don't really know the dynamo API - sorry, I've applied your changes.

arporter replied:
Ah, I see. I hadn't thought of that drawback before. Mind you, the chances are that nobody else is much the wiser either at this point (since if it hasn't got an interface description then we haven't touched it in quite a while).

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3615,51 +3608,48 @@ def gen_arg_setter_code(self, parent):
-        :param mapping: dictionary of access types (here INC) associated \
-                        with arguments with their metadata strings as keys
-        :type mapping: dict
-        :returns: a Fortran argument name.
```
Please put back the `:returns:` line in the docstring.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3615,51 +3608,48 @@ def gen_arg_setter_code(self, parent):
                                  "{1} access".
-                                 format(self.name, mapping["inc"]))
+                                 format(self.name,
+                                        rev_access_mapping[AccessType.INC]))
```
I've definitely not understood things yet. What's the difference between using `rev_access_mapping` and `api_name`? I'd naively have thought that they gave you the same information?

hiker replied:
No, you are spot on ... just that code was written before I decided to add the convenience function api_(specific)_name. I've updated this to use api_name() (and will change it once we agree on the final name for api_name()).

arporter replied:
Great!

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3615,51 +3608,48 @@ def gen_arg_setter_code(self, parent):
-                                        mapping["readwrite"]))
+                                 format(self.name,
+                                        rev_access_mapping[AccessType.WRITE],
+                                        rev_access_mapping
```
Strange line-break. Please could you adjust the formatting.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -3965,25 +3955,20 @@ def names(self):
-            if arg.access.lower() == mapping["write"] or \
-               arg.access.lower() == mapping["readwrite"] or \
-               arg.access.lower() == mapping["inc"]:
+            if arg.access == AccessType.WRITE or \
```
Could just do something like:

    if arg.access not in AccessType.write_accesses():

perhaps?

hiker replied:
OK, now I had that problem in several places: why is 'sum' missing in that list? AccessType has a 'all_write_access()' function, but it includes the reductions (i.e. sum). If we can add SUM here, then it's easy. If not, we need to find better names, which might get quite confusing ('all_write_access', 'all_write_non_reduction_accesses', ...).

arporter replied:
Off the top of my head I don't know the answer. Perhaps @rupertford can remember? I suspect we could put SUM in the list and see what happens...

hiker replied:
I started to have a look at this. From the docs it would indeed appear that a reduction should not be in that list (it is used to find a field to iterate over - in case of a SUM I'd guess the argument would be a scalar). So for now I am using existing AccessType functions:
```
            if arg.access in AccessType.all_write_accesses() and \
                    arg.access not in AccessType.get_valid_reduction_modes():
                return arg
```
I also tried to add a test case for this, but it turns out that dynamo3 does not use this function (it overwrites it with its own function and does not call the function in psyGen), and of course gocean does not support SUM in the first place. I'll just left it unchanged (except for the lines above, which I think will indeed increase portability later).

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4199,21 +4184,13 @@ def __init__(self, call, arg_info, access):
-        self._read_access_types = [MAPPING_ACCESSES["read"],
-                                   MAPPING_ACCESSES["readwrite"],
-                                   MAPPING_ACCESSES["inc"]]
+        self._write_access_types = [AccessType.WRITE,
```
Could these lists just be defined in AccessType?

hiker replied:
Sure (there already is 'all_read_access()'). Do you want to keep the self._XXXX_access_types variables, or should I replace them everywhere with corresponding functions in AccessType?

arporter replied:
I think it would be nice to have everything in one place (`AccessType`) if that works for you.

hiker replied:
OK, all _{read,write}_access_types are gone now :)

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4241,9 +4218,28 @@ def access(self):
-        ''' set the access type for this argument '''
+        '''Set the access type for this argument.
+        :param value: New access type.
+        :type value: :py:class:`psyclone.core.access_type`.
```
s/access_type/AccessType/

hiker replied:
See above :)

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4241,9 +4218,28 @@ def access(self):
+    def is_read(self):
+        '''Checks if the argument is read, i.e. has access
+        type read, readwrite, or inc.'''
+        return self._access in [AccessType.READ, AccessType.READWRITE,
```
`in self._read_access_types` perhaps?

hiker replied:
The two is_read/write functions are actually a left over before I decided to use enums. So I have removed the completely.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4241,9 +4218,28 @@ def access(self):
+    def is_written(self):
+        '''Checks if the argument is written, i.e. has access
+        type write, readwrite, or inc.'''
+        return self._access in [AccessType.WRITE, AccessType.READWRITE,
```
Again, `in self._write_access_types`?

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4241,9 +4218,28 @@ def access(self):
         self._access = value
 
+    def is_read(self):
+        '''Checks if the argument is read, i.e. has access
```
Please document the return value.

arporter commented on src/psyclone/psyGen.py:
```diff
@@ -4241,9 +4218,28 @@ def access(self):
+        return self._access in [AccessType.READ, AccessType.READWRITE,
+                                AccessType.INC]
+
+    def is_written(self):
```
Please document the return value.

arporter commented on src/psyclone/tests/conftest.py:
```diff
@@ -79,6 +79,28 @@ def have_graphviz():
+    # search path of the Configuration object will find the right
+    # config file. So only overwrite the default search path behaviour
+    # if the repository config file is actually found:
+    if os.path.isfile(config_file):
```
I love it when a plan comes together :-)

hiker commented on .travis.yml:
```diff
@@ -50,12 +50,14 @@ before_install:
+  - python -V
+  # python 2.7.14 on travis does not have enum, so in this case install
+  # a compatible enum implementation.
+  - if [ "$(python --version 2>&1)" == "Python 2.7.14" ]; then pip install enum34; fi
```
I admit I never looked at what setup.py does :) But agreed, and done - sees to work fine.

hiker commented on src/psyclone/configuration.py:
```diff
@@ -121,6 +121,18 @@ def get(do_not_load_file=False):
+        only exception).
+        '''
+        this_dir = os.path.dirname(os.path.abspath(__file__))
+        return os.path.join(this_dir, "..", "..", "config",
```
While I am not aware of anything not supporting ".." (and still running PSyclone), but using dirname twice indeed works.

hiker commented on src/psyclone/configuration.py:
```diff
@@ -335,7 +363,11 @@ def find_file():
 
+        The last entry is used for development when no installation was done.
+        It will then pick the psyclone.cfg file that is part of the
+        repository.
```
Hangover (and I don't say whose)

hiker commented on src/psyclone/configuration.py:
```diff
@@ -574,7 +606,104 @@ def get_default_keys(self):
+        if mapping:
+            from psyclone.core.access_type import AccessType
+            self._access_mapping = APISpecific.create_dict_from_string(mapping)
+            # Now convert the string type ("read" etc) to AccessType
```
Very good catch ... fixed, and I added a special test for this as well.

hiker commented on src/psyclone/core/access_type.py:
```diff
@@ -0,0 +1,125 @@
+        :rtype int'''
+        return len(AccessType)
+
+    def __str__(self):
```
__str__() returns (e.g. ) "INC", by default it would "AccessType.INC". I think it's worth keeping __str__(), I prefer it :)

arporter replied:
OK.

hiker commented on src/psyclone/core/tests/access_type_test.py:
```diff
@@ -0,0 +1,82 @@
+
+def test_str():
+    '''Tests conversion to an (API specific) string'''
+
```
Indeed, copy&paste error mixed the two test cases. Should be all fixed now.

hiker commented on src/psyclone/core/tests/access_type_test.py:
```diff
@@ -0,0 +1,82 @@
+
+def test_api_name():
+    '''Tests api_name(). '''
+    assert AccessType.INC.api_name() == "gh_inc"
```
Same copy&paste error - the setting of the config file was in the wrong test :(

hiker commented on src/psyclone/dynamo0p3.py:
```diff
@@ -582,20 +562,32 @@ def __init__(self, arg_type):
+        # Now convert from GH_* names to the generic access type:
+        api_config = Config.get().api_conf("dynamo0.3")
+        access_mapping = api_config.get_access_mapping()
+        rev_access_mapping = api_config.get_reverse_access_mapping()
```
No, it is used ... just way below. I've moved the assignment of rev_access_mapping to the location where it is actually used.

hiker commented on src/psyclone/dynamo0p3.py:
```diff
@@ -696,26 +688,28 @@ def __init__(self, arg_type):
-               VALID_REDUCTION_NAMES:
+            if self._access_descriptor.name not in [AccessType.READ] + \
+               AccessType.get_valid_reduction_modes():
+                api_name = rev_access_mapping[self._access_descriptor.name]
```
I really prefer to have 'api' in the name, since access_name() does not indicate that AccessType.INC.access_name() can be different depending on the API loaded. What about:
api_specific_name()  ?

arporter replied:
How about `api_access_name`?

hiker replied:
While I have to admit to prefer api_specific_name (the typical usage of this functions is:
```
  access.api_specific_name()   # vs:
  access.api_access_name()
```
But it doesn't really mattter, so api_access_name() it is now.

hiker replied:
If I understood you correct, you actually agreed to api_specific_name for the function, and then renaming the local variables. I've done that.

arporter replied:
Yes, I did - thanks!

hiker commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4203,25 +4200,39 @@ def __init__(self, alg_invocation, idx):
+        if access and not isinstance(access, AccessType):
+            api_config = Config.get().api_conf("dynamo0.3")
+            valid_names = api_config.get_valid_accesses_api()
+            raise GenerationError(
```
Done.

hiker replied:
Outdated, function removed. psyGen now raises InternalError in this case.

hiker commented:
@arporter : Than you very much for your effort in this review!!

I've also done all other corrections you mentioned (I thought I better save a dozen of emails just stating 'done' by just typing this once).

Outstanding before another review can be done:
1) Better name for api_name ... I suggested api_specific_name
2) Keep self._write_access_types etc, or replace their usage everywhere with an AccessType function call.

hiker commented on src/psyclone/dynamo0p3.py:
```diff
@@ -582,20 +562,32 @@ def __init__(self, arg_type):
+        rev_access_mapping = api_config.get_reverse_access_mapping()
+        try:
+            self._access_descriptor.name = \
+                access_mapping[self._access_descriptor.name]
```
Oops, I missed commenting on this point. Yes, I was surprised, too - and felt VERY bad when assigning an AccessType to the name. The problem is that _access_descriptor is a FunctionVar, which is used for the argument type:
```
    if arg_type.args[0].name not in GH_VALID_ARG_TYPE_NAMES:
...
            self._type = arg_type.args[0].name
...
        # The 2nd arg is an access descriptor
        self._access_descriptor = arg_type.args[1]
```
So I don't think that I can easily rename that attribute. I could add one more attribute perhaps, though ... it's odd to add this to a FunctionVar :(

The best option I see would be to create special classes for that would replace the FunctionVar there. I can't really estimate the scope of that, I am not deeply familiar with that code :( But I don't think it's a trivial change :(

hiker replied:
Dang - I was too eager. Turns out while the full FunctionVar is stored, only the name attribute of it was actually used, which made it quite easy to fix this properly (a199a225ac52828930dd1a02fb03e5184aa69fe1). Sorry for missing that.

arporter replied:
Oh good. This code will probably also be touched by @rupertford 's work on switching the meta-data parsing over to fparser2 (#239).

arporter commented:
@hiker, you're welcome :-) I think I've addressed the outstanding questions.

hiker commented:
I believe it is ready for another review.

arporter commented on src/psyclone/tests/gocean1p0_test.py:
```diff
@@ -1280,11 +1294,13 @@ def test06_kernel_invalid_access():
                            "test_files", "gocean1p0",
                            "test06_invoke_kernel_wrong_access.f90"),
               api="gocean1.0")
+    assert "compute_cu: argument access  is given as 'wrong' but must be one "\
```
Thanks for adding.

arporter commented on src/psyclone/tests/gocean1p0_transformations_test.py:
```diff
@@ -1421,7 +1428,12 @@ def test_go_loop_swap_errors():
                          "test_files", "dynamo0p3",
                          "1.0.1_single_named_invoke.f90"),
                     api="dynamo0.3")
+    # TODO: that still doesn't work, we have inconsistent values
```
Is this still true? If so it sounds like something that needs addressing...

hiker replied:
Done.

arporter commented on src/psyclone/tests/gocean1p0_transformations_test.py:
```diff
@@ -1421,7 +1428,12 @@ def test_go_loop_swap_errors():
                     api="dynamo0.3")
+    # TODO: that still doesn't work, we have inconsistent values
+    # in some psygen objects (read vs gh_read)
+    Config.get().api = "dynamo0.3"
```
Candidate for monkeypatching?

hiker replied:
Outdated.

arporter commented on src/psyclone/tests/kernel_transformation_test.py:
```diff
@@ -182,7 +182,7 @@ def test_new_kernel_file(tmpdir, monkeypatch):
     assert str(names[0]) == "continuity{0}_type".format(tag)
     # TODO #281 check compilation of code (needs Joerg's extension of
```
As you've touched this test, please could you add the compilation testing and remove this TODO?

hiker replied:
Done.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -864,7 +872,7 @@ def test_acc_dir_view(capsys):
     accdt = ACCEnterDataTrans()
     accpt = ACCParallelTrans()
-
+    Config.get().api = "gocean1.0"
```
monkeypatch?

hiker replied:
Outdated.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -914,6 +924,7 @@ def test_globalsum_view(capsys):
     then call view() on that.'''
     from psyclone.psyGen import colored, SCHEDULE_COLOUR_MAP
     from psyclone import dynamo0p3
+    Config.get().api = "dynamo0.3"
```
Is this line still needed despite the fixture?

hiker replied:
Done.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -1013,6 +1025,7 @@ def test_reduction_var_error():
 def test_reduction_sum_error():
     '''Check that we raise an exception if the reduction_sum_loop()
     method is provided with an incorrect type of argument'''
+    Config.get().api = "dynamo0.3"
```
Is this line needed?

hiker replied:
Nope :)

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -1999,13 +2013,15 @@ def test_node_is_valid_location():
 def test_node_ancestor():
     ''' Test the Node.ancestor() method '''
     from psyclone.psyGen import Loop
+    Config.get().api = "gocean1.0"
```
monkeypatch?

hiker replied:
Outdated.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2091,6 +2107,7 @@ def test_acc_dag_names():
     from psyclone.psyGen import ACCEnterDataDirective
     from psyclone.transformations import ACCEnterDataTrans, ACCParallelTrans, \
         ACCLoopTrans
+    Config.get().api = "gocean1.0"
```
monkeypatch?

hiker replied:
Outdated.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2556,6 +2574,7 @@ def test_find_w_args_multiple_deps(monkeypatch, annexed):
 def test_loop_props():
     ''' Tests for the properties of a Loop object. '''
     from psyclone.psyGen import Loop
+    Config.get().api = "gocean1.0"
```
monkeypatch/use fixture?

hiker replied:
Outdated.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2564,11 +2583,14 @@ def test_loop_props():
 def test_node_abstract_methods():
     ''' Tests that the abstract methods of the Node class raise appropriate
     errors. '''
+    Config.get().api = "gocean1.0"
```
monkeypatch/fixture?

hiker replied:
Outdated.

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2581,18 +2603,21 @@ def test_node_abstract_methods():
     ''' Test that we can obtain the fparser2 AST of a kernel. '''
     from psyclone.gocean1p0 import GOKern
     from fparser.two import Fortran2003
+    Config.get().api = "gocean1.0"
```
monkeypatch/fixture?

hiker replied:
@arporter , there are several places where I set a default api in a test function, mostly in 'generic' (non-api-specific) tests like alggen_test etc. By now in nearly all places this is not necessary for the tests anymore. Short explanation:
Originally I used Config.get().api_conf() in dynamo* and gocean*. For this to work I indeed needed to change the default API in the tests to make sure the right api-specific config will be picked up. But then I realised that in e.g. dynamo0p3 I should always get the dynamo0.3 config, so I changed the code to use: config.get().api_conf("dynamo0.3"). Since the API-files (dynamo*, gocean*) are responsible for translating the api-specific strings to the AccessType, there is no need to have a default API specified ... unless there is an error. In case of an error (especially if detected by a baseclass e.g. in psygen) a default API is needed to translate AccessType back to the api-specific string.

As summary: in all test cases that do not trigger an (AccessType-related) error, there is no need to change the API anymore, only the tests which test (AccessType-related) exception handling need a default being defined in order for the right error message.

So I could basically remove nearly all settings of an API.  The fixture in each API-file still makes sense (since e.g. a gocean1p0 test will likely test gocean1p0 related errors).

So would you be ok if I just remove the unnecessary settings of the api (with the only exception being tests that raise accesstype-related errors) - and since we have a fixture, that would only leave some non-API-specific tests (psyGen_test, alggen_test) where it is actually required.

arporter replied:
Yes, that sounds good to me!

hiker replied:
OK, I changed this everywhere, I've added 'outdated' to all the corresponding comments here.

hiker replied:
Done.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -6528,13 +6525,13 @@ def gen_stub(self):
         ''' Returns the argument corresponding to a field or operator that has
         WRITE access '''
-        return Kern.written_arg(self, FIELD_ACCESS_MAP)
+        return Kern.written_arg(self)
```
This line is not covered but, since all we're doing is calling the `written_arg` method of the parent class, this method can be removed altogether. In fact, the same applies for `incremented_arg` above.

hiker replied:
Done (except that those functions here had @property, which psyGen did not have ... but since I don't like @property anyway, I've added '()' to all accesses of the former property).

TeranIvy commented on src/psyclone/dynamo0p3.py:
```diff
@@ -646,31 +638,31 @@ def __init__(self, arg_type):
             # TODO: extend for "gh_write"
             if self._function_space1.lower() in CONTINUOUS_FUNCTION_SPACES \
-               and self._access_descriptor.name.lower() == "gh_readwrite":
+               and self._access_descriptor.name == AccessType.READWRITE:
```
It did not, in fact we discourage using it, see issue #138.

hiker commented on src/psyclone/configuration.py:
```diff
@@ -121,6 +121,18 @@ def get(do_not_load_file=False):
+        '''This function returns the absolute path to the config file included
+        in the PSyclone repository. It is used by the testing framework to make
+        sure all tests get the same config file (see tests/config_tests for the
+        only exception).
```
Done.

hiker commented on src/psyclone/configuration.py:
```diff
@@ -574,7 +606,104 @@ def get_default_keys(self):
+class APISpecific(object):
+    '''A base class for functions that each API-specific class must provide.
+    At the moment that is a function access_mapping().
+    :param section: :py:class:`configparser.SectionProxy`
```
Done.

hiker commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4203,25 +4200,39 @@ def __init__(self, alg_invocation, idx):
                     global_sum = DynGlobalSum(scalar, parent=loop.parent)
                     loop.parent.children.insert(loop.position+1, global_sum)
 
+    def unique_declarations(self, datatype, access=None):
```
Outdated, the whole function in dynamo0p3 has been deleted, it is all handled in psyGen now.

hiker commented on src/psyclone/configuration.py:
```diff
@@ -574,7 +606,106 @@ def get_default_keys(self):
+        :returns: A dictionary with the key,value pairs from the input string.
+        :rtype: dict.
+        :Raises ConfigurationError if the input string contains an entry
+                that is does not have a ":".
```
Done.

arporter replied:
Almost! You need a colon after "ConfigurationError" :-)

hiker replied:
Gee, that was embarrassing (then again ... good case for being able to create an API guide, since it would allow a dev to easily check the correctness of markups). Done.

hiker commented on src/psyclone/configuration.py:
```diff
@@ -574,7 +606,106 @@ def get_default_keys(self):
+        return self._reverse_access_mapping
+
+    def get_valid_accesses_api(self):
+        '''Returns the sorted, API-specific sorted names of all valid access
```
Done

hiker commented on src/psyclone/configuration.py:
```diff
@@ -574,7 +606,106 @@ def get_default_keys(self):
+# =============================================================================
+class APISpecificConfig(object):
+    '''A base class for functions that each API-specific class must provide.
+    At the moment that is a function access_mapping().
```
Done.

hiker commented on src/psyclone/dynamo0p3.py:
```diff
@@ -696,26 +684,29 @@ def __init__(self, arg_type):
+            if self._access_type not in [AccessType.READ] + \
+               AccessType.get_valid_reduction_modes():
+                rev_access_mapping = api_config.get_reverse_access_mapping()
+                api_name = rev_access_mapping[self._access_type]
```
Done (and as indicated previously, I also renamed the api_name function in AccessType).

arporter replied:
Great - thank you.

hiker commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4201,25 +4196,28 @@ def __init__(self, alg_invocation, idx):
+                format(str(GH_VALID_ARG_TYPE_NAMES), datatype))
+        if access and not isinstance(access, AccessType):
+            api_config = Config.get().api_conf("dynamo0.3")
+            valid_names = api_config.get_valid_accesses_api()
```
Done (in psyGen, this function in dynamo03 has been removed).

hiker commented on src/psyclone/psyGen.py:
```diff
@@ -2278,13 +2279,16 @@ def view(self, indent=0):
         '''Return the name of all scalars within this region that require a
-        reduction of type reduction_type. Returned names will be unique. '''
+        reduction of type reduction_type. Returned names will be unique.
+        :param: reduction_type The reduction type (e.g. AccessType.SUM) to
```
Done.

hiker commented on src/psyclone/dynamo0p3.py:
```diff
@@ -4201,25 +4196,28 @@ def __init__(self, alg_invocation, idx):
         specified datatype.  If access is supplied (e.g. "gh_write")
         then only declarations with that access are returned. '''
-        if datatype not in VALID_ARG_TYPE_NAMES:
+        if datatype not in GH_VALID_ARG_TYPE_NAMES:
```
Done (outdated)

hiker commented on src/psyclone/core/tests/access_type_test.py:
```diff
@@ -0,0 +1,80 @@
+
+
+def test_api_name():
+    '''Tests api_name(), i.e. conversion to an
```
Done.

arporter replied:
Great. Thanks for taking the trouble to change it all again.

hiker commented on src/psyclone/tests/alggen_test.py:
```diff
@@ -411,25 +419,31 @@ def test_single_invoke_dynamo0p1():
         specified in an invoke call for the dynamo0.1 API. We use the
         generate function as parse and PSyFactory need to be called before
         AlgGen so it is simpler to use this. '''
+    Config.get().api = "dynamo0.1"
```
As discussed elsewhere: it turns out in nearly all cases setting the api is not necessary anymore (since it is actually only used in case of errors), so while I had started using monkeypatch (and finally started to understand what it does ;) ), I could remove all those assignments. It turns out that the error testing is done in functions that set the right API in a fixture.

arporter replied:
I'm a bit confused here - I see that you've changed the test to set the API using monkeypatch BUT it's being set to 'gocean1.0' (lines 422 and 438). I guess that the errors that those tests are provoking happen before the kernel meta-data is parsed so we get away with it?

arporter replied:
Having got through to the end of all the comments I think I finally understand what's going on :-) Are this line and the one I comment on below just hangovers that should be removed?

hiker replied:
Yes, left over ... and I am wildly cursing here: I grep'ed to find all settings of the api that I had added to the tests, but used:
```
grep "Config.get().api" *py
```
but the monkeypatched lines were not detected by the grep :( Note to self: `grep "Config.get().*api"` :)

arporter replied:
:-)

hiker commented on src/psyclone/tests/alggen_test.py:
```diff
@@ -411,25 +419,31 @@ def test_single_invoke_dynamo0p1():
         not contain any actual invoke() calls. We use the generate
         function as parse and PSyFactory need to be called before
         AlgGen so it is simpler to use this. '''
+    Config.get().api = "dynamo0.1"
```
Outdated.

arporter replied:
See previous comment - it's being set to "gocean1.0" instead of "dynamo0.1"

hiker replied:
Removed.

hiker commented on src/psyclone/tests/config_test.py:
```diff
@@ -117,6 +121,13 @@ def int_entry(request):
 
 
+def test_get_repo_config_file():
+    '''Check that we can find the expected config file in the repository.
```
Done.

hiker commented on src/psyclone/tests/config_test.py:
```diff
@@ -491,6 +511,61 @@ def test_incl_path_errors(tmpdir):
+        APISpecificConfig.create_dict_from_string("k1:v1, k2:v2:something")
+    assert mapping == {"k1": "v1", "k2": "v2:something"}
+
+    # Tests errors: '=' is not valid, will
```
Fixed.

hiker commented on doc/configuration.rst:
```diff
@@ -113,6 +121,33 @@ REPROD_PAD_SIZE         If generating code for reproducible OpenMP reductions,
+                        gh_inc:inc, gh_sum:sum``
+
+                        At this stage these 5 types are defined for read, write,
+                        read+write, increment and summation access by PSyclone.
```
Done.

hiker commented on src/psyclone/tests/dynamo0p1_transformations_test.py:
```diff
@@ -10,13 +10,21 @@
 
+@pytest.fixture(scope="module", autouse=True)
+def setup():
+    '''Make sure that all tests here use dynamo0.3 as API.'''
```
Done.

hiker commented on src/psyclone/tests/dynamo0p3_lma_test.py:
```diff
@@ -42,6 +42,7 @@
 import os
 import pytest
 from fparser import api as fpapi
+from psyclone.core.access_type import AccessType
```
Done.

hiker commented on src/psyclone/tests/dynamo0p3_stubgen_test.py:
```diff
@@ -98,12 +98,12 @@ def test_dynscalars_err(monkeypatch):
     arg._type = "invalid-scalar-type"
     # Monkeypatch the list of supported scalar types to include this one
-    monkeypatch.setattr(dynamo0p3, "VALID_SCALAR_NAMES",
+    monkeypatch.setattr(dynamo0p3, "GH_VALID_SCALAR_NAMES",
```
Done.

hiker commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1807,13 +1814,27 @@ def test_invoke_uniq_declns_invalid_access():
 
+def test_invoke_uniq_declns_valid_access():
+    ''' tests that valid access modes ('read', 'write') are
+    accepted by  Invoke.unique_declarations() is called.'''
```
Done.

hiker commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1807,13 +1814,27 @@ def test_invoke_uniq_declns_invalid_access():
+                                        "1.7_single_invoke_2scalar.f90"),
+                           api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=True).create(invoke_info)
+    print("type is", type(psy.invokes.invoke_list[0]))
```
Done.

hiker commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -1807,13 +1814,27 @@ def test_invoke_uniq_declns_invalid_access():
+                           api=TEST_API)
+    psy = PSyFactory(TEST_API, distributed_memory=True).create(invoke_info)
+    print("type is", type(psy.invokes.invoke_list[0]))
+    psy.invokes.invoke_list[0].unique_declarations("gh_field",
```
Done.

hiker commented on src/psyclone/tests/dynamo0p3_test.py:
```diff
@@ -3018,17 +3027,17 @@ def test_arg_descriptor_init_error():
     # Now try to trip the error by making the initial test think
     # that GH_INVALID is actually valid
-    from psyclone.dynamo0p3 import VALID_ARG_TYPE_NAMES, DynArgDescriptor03
+    from psyclone.dynamo0p3 import GH_VALID_ARG_TYPE_NAMES, DynArgDescriptor03
```
Done.

hiker commented on src/psyclone/tests/gocean0p1_test.py:
```diff
@@ -39,11 +39,12 @@
-from fparser.api import parse as parse1
 
 API = "gocean0.1"
 
```
Done.

hiker commented on src/psyclone/tests/gocean1p0_test.py:
```diff
@@ -1114,6 +1122,12 @@ def test_raw_arg_list_error(monkeypatch):
         _ = kern.arguments.raw_arg_list()
     assert ("Kernel compute_z_code, argument z_fld has unrecognised type: "
             "'broken'" in str(err))
+    # Also test that assigning a non-AccessType value to a kernel argument
```
Done.
The strange error message is caused by a difference in output between python2 and 3. Similar tests exist elsewhere (psyGen_test.py: 'assert "Unrecognized attribute type " in str(error.value)'), but I have replaced this test with a regular expression tests that also checks for 'str' in the output (i.e. independent of 'class' (python3 iirc) or 'type' (python2) in the string).

arporter replied:
Great, thanks. Picky, but could you remove the "Also" from the comment on line 1136?

hiker replied:
Done.

hiker commented:
Codecov/patch still shows only 99.13% target hit ... but I have to admit I can't figure out how to see which new lines were not hit. I see dynamo0p3 in red, but when I click on the red rectangle, I get: "File/folder was not found on base commit. No compare to display. " And for dynamo0.3 I see a +0.07% - in short: I have no idea how to figure out what might be wrong there. So I'll mark this as ready for review, but if there is something that I missed, please enlighten me how I can test this myself.

arporter commented:
All tests, including compilation, pass.

re. Coverage, what you want is the 'Diff' tab on CodeCov. This then shows (if you scroll down) that the code that is missed is the `updated_arg` method in dynamop3.py. It looks as though this is never called which is a bit suspicious. Maybe it's completely unnecessary?

arporter commented on src/psyclone/dynamo0p1.py:
```diff
@@ -379,6 +380,7 @@ def gen_code(self, parent):
                        position=["before", position])
 
 
+
```
pycodestyle now complains that there are too many blank lines.

hiker replied:
Done.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -734,24 +726,24 @@ def function_space_to(self):
         the first function space specified in the metadata. Raise an
         error if this is not an operator. '''
-        if self._type in VALID_OPERATOR_NAMES:
+        if self._type in GH_VALID_OPERATOR_NAMES:
```
Not strictly modified by you but since you've altered the `if`, please could you fix pylint's complaint about the unnecessary `else` below.

hiker replied:
Done. I admit I noticed those warnings, but I usually prefer to have patches as 'clean' as possible, so that it is easy for the reviewer to see the important changes. I was thinking of a separate pull request that improves coding style ;) But, if you want it, you get it :)  Done.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -734,24 +726,24 @@ def function_space_to(self):
         the second function space specified in the metadata. Raise an
         error if this is not an operator. '''
-        if self._type in VALID_OPERATOR_NAMES:
+        if self._type in GH_VALID_OPERATOR_NAMES:
```
Again, pylint doesn't like the `else` below.

hiker replied:
Done.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -760,9 +752,9 @@ def function_space(self):
         if self._type == "gh_field":
             return self._function_space1
-        elif self._type in VALID_OPERATOR_NAMES:
+        elif self._type in GH_VALID_OPERATOR_NAMES:
```
This can just be `if ...` and same again below.

hiker replied:
Done.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -776,10 +768,10 @@ def function_spaces(self):
         if self._type == "gh_field":
             return [self.function_space]
-        elif self._type in VALID_OPERATOR_NAMES:
+        elif self._type in GH_VALID_OPERATOR_NAMES:
```
Again, this can just be `if` and same again below.

hiker replied:
Done.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -5790,11 +5800,7 @@ def _upper_bound_fortran(self):
-        else:
-            my_mapping = FIELD_ACCESS_MAP
-        return Loop.has_inc_arg(self, my_mapping)
+        return Loop.has_inc_arg(self)
```
Can we get rid of the `mapping` argument to `has_inc_arg` now?

hiker replied:
I removed the whole function, which is not necessary anymore since it only calls the function in the base class.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -5830,18 +5836,18 @@ def _halo_read_access(self, arg):
                     "accesses. Found '{0}'.".format(self._upper_bound_name))
             return self._upper_bound_name in ["cell_halo", "ncells"]
-        if arg.type in VALID_SCALAR_NAMES:
+        if arg.type in GH_VALID_SCALAR_NAMES:
```
Please replace the `elif`s that are after `return`s with just `if`

hiker replied:
Done.

arporter commented on src/psyclone/dynamo0p3.py:
```diff
@@ -8495,19 +8491,21 @@ def intent(self):
         :rtype: str
         '''
-        if self.access == "gh_read":
+        if self.access == AccessType.READ:
```
Please simplify the `elif`s to `if`s where possible to keepy pylint happy.

hiker replied:
Done.

arporter commented on src/psyclone/configuration.py:
```diff
@@ -574,7 +606,108 @@ def get_default_keys(self):
+class APISpecificConfig(object):
+    '''A base class for functions that each API-specific class must provide.
+    At the moment this is just the function 'access_mapping' that maps between
+    API-specific access-descriptor strings and the PSyclone internal 
```
pylint complains about trailing white space here.

hiker replied:
And I just learned that if you use pycharm, and 'go to the end of a line' ... it goes to the last non-space character in a line (i.e. before any trailing white-space) ... strange. Anyway, fixed :)

arporter replied:
I don't even know what pycharm is :-)

arporter commented on src/psyclone/tests/psyGen_test.py:
```diff
@@ -2569,6 +2579,7 @@ def test_loop_props():
 def test_node_abstract_methods():
     ''' Tests that the abstract methods of the Node class raise appropriate
     errors. '''
+    from psyclone.psyGen import Node
```
pylint complains that `Node` has already been imported.

hiker replied:
Fixed.

arporter commented:
There are a handful of pylint errors to address (most of them aren't strictly yours) and a few remaining threads from previous reviews that we need to resolve. Once that's done then this can be merged.

hiker commented:
> the code that is missed is the `updated_arg` method in dynamop3.py. It looks as though this is never called which is a bit suspicious. Maybe it's completely unnecessary?

Yes, I agree that it appears to be unused, so I removed it which then made another function unused (except for in tests, so I removed the tests as well).

hiker commented:
> re. Coverage, what you want is the 'Diff' tab on CodeCov. 
OK, that is good to know ... though as far as I can tell it still means you have to manually scroll through hundreds of lines, trying to spot the ones that are red (which are often and have a "+" (i.e. were added/modified in the patch). But ok, I learned something :) I wonder if it would be worth to add some of the details to either the dev-guide, or the wiki (code review part) at least.

hiker commented:
Ready to go again.

arporter commented:
I'll merge this once the final Travis checks have completed.
